### DISCUSSION TITLE
How to include LaTeX macros in YAML header for html and revealjs documents?

### DISCUSSION THREAD

**Discussion by mpf** [quarto-dev/quarto-cli#8376] [Q&A] [answered]
### Description

This question is a follow-up to an earlier [post](https://github.com/quarto-dev/quarto-cli/discussions/2845), which didn't seem to resolve. I wish to include latex macros into html and revealjs formats via the yaml header. But I'm not able to get this to work because the macros are printed at the top of the output. (See below.) Here's a working example with illustrates the behavior. 

`slide.qmd` file:
````qmd
---
title: Test
format:
    revealjs:
      include-before-body: _macros.qmd
---

## First slide

The nonnegative orthant: $\Rn_+$
````

`_macros.qmd` file:
````qmd
::: {.content-hidden}
$$
\def\Rn{\mathbb{R}^n}
$$
:::
````
And here's the output:

<img width="1043" alt="Screenshot 2024-01-21 at 2 48 54 PM" src="https://github.com/quarto-dev/quarto-cli/assets/109942/23aa6634-9f0a-4aeb-a739-4cd464a9e946">

**Comment by mcanouil** [comment]
That's the way to go in general in Quarto.
For specifics, go to the Math library documentation website, by default, it is MathJax.
In other words, if the macro you defined is not supported by MathJax (or the chosen library), there is absolutely nothing Quarto can do about that.

(This is also part of the thread you linked.)

**Comment by mpf** [comment]
It's not the macro that's incorrect — it's the method of inclusion via the yaml header that I'm trying to get right. Here's an example that shows how the macro is correctly rendered. But that only works if it's included in the body, which is inconvenient, because it then has to be included in *every* source file.

````qmd
---
title: Test
format: revealjs
# format:
#     revealjs:
#       include-before-body: _macros.qmd
---

## First slide

{{< include _macros.qmd >}}

The nonnegative orthant: $\Rn_+$
````

`_macros.qmd`:
````qmd
$$
\def\Rn{\mathbb{R}^n}
$$
````
Generated output, which is correct:

<img width="993" alt="Screenshot 2024-01-21 at 4 34 21 PM" src="https://github.com/quarto-dev/quarto-cli/assets/109942/3a8fd649-1562-449f-b280-2e67319ea314">

**Reply by mcanouil** [reply]
I see, but that's MathJax 🤷‍♂️
If it requires the macro to be in the body, that's not up to Quarto.

You could look in MathJax documentation for another way to provide the macros such as from the configuration JavaScript function. The configuration can be done outside of body.
But again, that's MathJax not Quarto (there are several discussions about include custom
MathJax settings already).

**Comment by cderv** [comment]
@mpf as documented in https://quarto.org/docs/output-formats/html-basics.html#includes, these includes are inserted **verbatim** in the document. This means they are not processed, so you can't use a .qmd and expect it to be rendered before being included. 

This is why `include-before-body: _macros.qmd` results in the file content being shown as is in the slides, because it will be inserted and considered as HTML content directly. 

Using `include` shortcode will insert the content of the file before rendering so it will be rendered as you expect. And that is why it works. 

So this has nothing to do with Mathjax syntax, its support, or any expectation to be in body. Here, you are just using `include-before-body` incorrectly. Using the `{{< include _macros.qmd >}}` is a good way to do that IMO. 

Hope it helps understand.

**Comment by mpf** [comment, marked as answer]
Using `{{< include _macros.qmd >}}` works but has the unfortunate side effect that, if placed before the first header in a revealjs format document, it can cause a blank slide. That means having to place it inside of the first slide. That's awkward.

I now use this approach: in the yaml header that specifies the format, I have

```yaml
format:
  revealjs:
    include-in-header:
      - file: "macros.jax"
```
where `macros.jax` contains the macro definitions, as specified [here](https://docs.mathjax.org/en/latest/input/tex/macros.html):
```
<script>
window.MathJax = {
  tex: {
    macros: {
      RR: "{\\bf R}",
      bold: ["{\\bf #1}", 1]
    }
  }
};
</script>
```
This seems to work reliably, though the downside is having to maintain macros in an unusual format.

**Reply by cderv** [reply]
> but has the unfortunate side effect that, if placed before the first header in a revealjs format document, it can cause a blank slide. That means having to place it inside of the first slide. That's awkward.

About this, you can also hide a slide from the presentation: 
https://quarto.org/docs/presentations/revealjs/advanced.html#slide-visibility

Could be useful in such case.
